common/src/main/java/org/apache/drill/common/scanner/ClassPathScanner.java

```diff
@@ -0,0 +1,458 @@
+
+import javassist.bytecode.AccessFlag;
+import javassist.bytecode.AnnotationsAttribute;
+import javassist.bytecode.ClassFile;
+import javassist.bytecode.FieldInfo;
+import javassist.bytecode.annotation.AnnotationMemberValue;
+import javassist.bytecode.annotation.ArrayMemberValue;
+import javassist.bytecode.annotation.BooleanMemberValue;
+import javassist.bytecode.annotation.ByteMemberValue;
+import javassist.bytecode.annotation.CharMemberValue;
+import javassist.bytecode.annotation.ClassMemberValue;
+import javassist.bytecode.annotation.DoubleMemberValue;
+import javassist.bytecode.annotation.EnumMemberValue;
+import javassist.bytecode.annotation.FloatMemberValue;
+import javassist.bytecode.annotation.IntegerMemberValue;
+import javassist.bytecode.annotation.LongMemberValue;
+import javassist.bytecode.annotation.MemberValue;
+import javassist.bytecode.annotation.MemberValueVisitor;
+import javassist.bytecode.annotation.ShortMemberValue;
+import javassist.bytecode.annotation.StringMemberValue;
+
+/**
+ * Classpath scanning utility.
+ * The classpath should be scanned once at startup from a DrillConfig instance. {@see ClassPathScanner#fromPrescan(DrillConfig)}
+ * The DrillConfig provides the list of packages to scan. (drill.classpath.scanning.packages) {@see CommonConstants#IMPLEMENTATIONS_SCAN_PACKAGES}
+ * Only the class directories and jars containing a drill-module.conf will be scanned.
+ * Drill core packages are scanned at build time and the result is saved in a JSON file. {@see ClassPathScanner#FUNCTION_REGISTRY_FILE}
+ * At runtime only the packages that have not be scanned yet will be scanned.
+ */
+public final class ClassPathScanner {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ClassPathScanner.class);
+  private static final JavassistAdapter METADATA_ADAPTER = new JavassistAdapter();
+
+  /**
+   * defines all the things to scan for
+   */
+  private static final Set<String> SCANNED_CLASSES = unmodifiableSet(
+      new HashSet<>(asList(
+          "org.apache.drill.common.logical.data.LogicalOperator",
+          "org.apache.drill.common.logical.FormatPluginConfig",
+          "org.apache.drill.common.logical.StoragePluginConfig",
+          "org.apache.drill.exec.expr.DrillFunc",
+          "org.apache.drill.exec.expr.fn.PluggableFunctionRegistry",
+          "org.apache.drill.exec.physical.base.PhysicalOperator",
+          "org.apache.drill.exec.physical.impl.BatchCreator",
+          "org.apache.drill.exec.physical.impl.RootCreator",
+          "org.apache.drill.exec.rpc.user.security.UserAuthenticator",
+          "org.apache.drill.exec.store.dfs.FormatPlugin",
+          "org.apache.drill.exec.store.StoragePlugin",
+          "org.apache.hadoop.hive.ql.udf.generic.GenericUDF",
+          "org.apache.hadoop.hive.ql.exec.UDF"
+          )));
+
+
+  /**
+   * scans the inheritance tree
+   */
+  private static class SubTypesScanner extends AbstractScanner {
```

Comment:
Can you explain why we implemented our own when I believe a SubTypesScanner exists in the reflections tree?